mtags/src/main/scala-2/scala/meta/internal/pc/ScalaPresentationCompiler.scala

```diff
@@ -118,6 +125,24 @@ case class ScalaPresentationCompiler(
 
   def didClose(uri: URI): Unit = {}
 
+  override def semanticTokens(
+      params: VirtualFileParams,
+      capableTypes: util.List[String],
+      capableModifiers: util.List[String]
+  ): CompletableFuture[ju.List[Integer]] = {
```

Comment:
What do you think about returning the list of the data structure in the form of

```typescript
{ line: 2, startChar:  5, length: 3, tokenType: "property",
	tokenModifiers: ["private", "static"]
},
{ line: 2, startChar: 10, length: 4, tokenType: "type", tokenModifiers: [] },
{ line: 5, startChar:  2, length: 7, tokenType: "class", tokenModifiers: [] }
```

(copied from https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens)

which makes

- SemanticHighlightingProvider much simpler
  - it looks like, currently, we directly compute the integer encoded representation from the typed AST. That is I believe hard to debug and read.
  - Transforming the data step by step makes things easier to debug.
- It will be easier to test
  - If the presentation compiler returns a more meaningful data structure (instead of integer decoded data), we can use them directly from `TestingServer` instead of encoding back the integers.

and transform them into integers representation later?

Reply:
I have enough time to do that, so I can't .
But I will note the specifications as comment.